### PR TITLE
Exclude tags in search, add "hide DLC mods" checkbox

### PR DESCRIPTION
**Pull request by xen-42**
image

Adds `exclude-tag` url param. 
One click makes all results include the tag, click again to exclude it, click again to reset back to normal.

If we implement this we should have an update on the manager at the same time that also implements tag excluding.

**EDIT:**

Edited to add "Hide DLC mods" checkbox that sets the tag to exclude immediately to make it more clear. Leaving in excluding tags because they could still be useful, even if they are a bit difficult to find.

Also made the excluded mods have a strike-through to make it extra clear what it means (and for colourblindness) after talking with Rai

image

### CODE REVIEW

**Comment by xen-42** [comment]
Clear button is broken right now, still need to fix some stuff

**Comment by xen-42** [comment]
Fixed clear button, only issue is making onChange optional on CheckboxInput

**Comment by xen-42** [comment]
Okay everything works now as far as I can tell, ready for review

File: src/lib/components/checkbox-input.svelte [hunk 1]
```diff
@@ -1,8 +1,9 @@
 <script lang="ts">
 	export let checked: boolean;
+	export let onChange: (event : any) => void = () => {};
 </script>
 
 <label class="input px-2 cursor-pointer flex flex-row items-center gap-1">
-	<input type="checkbox" class="accent-accent bg-accent text-white" bind:checked />
+	<input type="checkbox" class="accent-accent bg-accent text-white" bind:checked on:change={onChange} />
```
Comment (Raicuparta) [inline src/lib/components/checkbox-input.svelte]: if you just do `<input on:change />` the event will be forwarded so you can do `<CheckboxInput on:change={etc} />`. The TagToggle component has an example of this.

Also `any` is evil.

**Review by Raicuparta — CHANGES_REQUESTED** [review]

File: src/lib/components/mod-grid/mod-grid.svelte [hunk 2]
```diff
@@ -1,5 +1,8 @@
 <script context="module" lang="ts">
-	export type TagStates = Record<string, boolean>;
+	export type TagStates = Record<string, string>;
+	export const tagIncluded = 'included'
+	export const tagExcluded = 'excluded'
+	export const dlcTag = 'requires-dlc'
```
Comment (Raicuparta) [inline src/lib/components/mod-grid/mod-grid.svelte:5]: maybe put these in constants.ts

File: src/lib/components/mod-grid/mod-grid.svelte [hunk 3]
```diff
@@ -109,26 +123,54 @@
 	};
 
 	const onToggleTag = (tag: string) => {
-		const { [tag]: toggledTag, ...currentTagStates } = tagStates;
+		let toggledTag = tagStates[tag];
 
-		if (!toggledTag) {
-			currentTagStates[tag] = true;
+		if (toggledTag == undefined) {
```
Comment (Raicuparta) [inline src/lib/components/mod-grid/mod-grid.svelte]: `===`

File: src/lib/components/mod-grid/mod-grid.svelte [hunk 4]
```diff
@@ -109,26 +123,54 @@
...
+			setTagState(tag, tagIncluded);
 		}
+		else if (toggledTag == tagIncluded) {
```
Comment (Raicuparta) [inline src/lib/components/mod-grid/mod-grid.svelte]: `===`

File: src/lib/components/mod-grid/mod-grid.svelte [hunk 5]
```diff
@@ -167,6 +209,12 @@
 		<div>
 			<CheckboxInput bind:checked={showDetails}>Show details</CheckboxInput>
 		</div>
+		<div>
+			<CheckboxInput bind:checked={hideDLC} onChange={(event) => {
+				if (!event || !event.currentTarget) return;
```
Comment (Raicuparta) [inline src/lib/components/mod-grid/mod-grid.svelte]: pointless check?

File: src/lib/components/mod-grid/mod-grid.svelte [hunk 6]
```diff
@@ -1,5 +1,8 @@
 <script context="module" lang="ts">
-	export type TagStates = Record<string, boolean>;
+	export type TagStates = Record<string, string>;
+	export const tagIncluded = 'included'
+	export const tagExcluded = 'excluded'
```
Comment (Raicuparta) [inline src/lib/components/mod-grid/mod-grid.svelte]: this should be a typed string instead of consts

Comment (Raicuparta) [reply src/lib/components/mod-grid/mod-grid.svelte]: or enum

File: src/lib/components/checkbox-input.svelte (hunk 1 above)
Comment (xen-42) [reply src/lib/components/checkbox-input.svelte]: Forwarded it instead

File: src/lib/components/mod-grid/mod-grid.svelte (hunk 4 above)
Comment (xen-42) [reply src/lib/components/mod-grid/mod-grid.svelte]: Replaced all my == with ===

File: src/lib/components/mod-grid/mod-grid.svelte (hunk 5 above)
Comment (xen-42) [reply src/lib/components/mod-grid/mod-grid.svelte]: Yeah I just copied that from elsewhere didnt even know what it did, got rid of it

**Review by JohnCorby — APPROVED** [review]

**Pull request merged**